bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java

```diff
@@ -0,0 +1,546 @@
+                } else if (service.equals(HUB_SERVICE_LIST_ZONE)) {
+                    processMsgListZone(hubMsg);
+                } else if (service.equals(HUB_SERVICE_SYSTEM_INFO)) {
+                    processMsgSystemInfo(hubMsg);
+                }
+            }
+        } catch (Exception e) {
+            logger.error("Hub controller failed ({})", e.getMessage());
+        } finally {
+            synchronized (this) {
+                // If there are still pending requests we need to cancel them
+                for (AdorneHubController.ZoneIDFutureRecord stateCommand : stateCommands) {
+                    stateCommand.future.cancel(false);
+                }
+                for (CompletableFuture<List<Integer>> zoneCommand : zoneCommands) {
+                    zoneCommand.cancel(false);
+                }
+                for (CompletableFuture<String> macAddressCommand : macAddressCommands) {
+                    macAddressCommand.cancel(false);
+                }
+                CompletableFuture<@Nullable Void> hubControllerConnected = this.hubControllerConnected;
+                hubControllerConnected.cancel(false);
+            }
+            disconnect();
+            logger.info("Exiting hub controller");
+        }
+    }
+
+    private synchronized boolean shouldStop() {
+        boolean timeoutExceeded = stopTimestamp > 0 && System.currentTimeMillis() > stopTimestamp;
+        boolean commandsServed = stopWhenCommandsServed.get() && stateCommands.isEmpty() && zoneCommands.isEmpty()
+                && macAddressCommands.isEmpty();
+
+        return (stop.get() || timeoutExceeded || commandsServed);
+    }
+
+    private boolean connect() {
+        try {
+            synchronized (this) {
+                if (hubSocket == null) {
+                    Socket hubSocket = this.hubSocket = new Socket(hubHost, hubPort);
+                    hubSocket.setSoTimeout(HUB_CONNECT_TIMEOUT);
+                    hubOut = new PrintStream(hubSocket.getOutputStream());
+                    Scanner hubIn = this.hubIn = new Scanner(hubSocket.getInputStream());
+                    hubIn.useDelimiter("\0");
+                    logger.debug("Hub connection established");
+
+                    CompletableFuture<@Nullable Void> hubControllerConnected = this.hubControllerConnected;
+                    hubControllerConnected.complete(null);
+
+                    AdorneHubChangeNotify changeListener = this.changeListener;
+                    if (changeListener != null) {
+                        changeListener.connectionChangeNotify(true);
+                    }
+                }
+            }
+            return true;
+        } catch (IOException e) {
+            try {
+                logger.debug("Couldn't establish hub connection ({}). Sleep {} ms before re-trying.", e.getMessage(),
+                        hubReconnectSleep);
+                Thread.sleep(hubReconnectSleep);
```

Comment:
Don't use `sleep`. Schedule the next execution with `scheduler.schedule(...)`

Reply:
Done